include/onnxruntime/core/framework/stream_handles.h

```diff
@@ -0,0 +1,159 @@
+
+  StreamHandle GetHandle() const { return handle_; }
+
+  const OrtDevice& GetDevice() const { return device_; }
+
+  // We use the timestamp based vector clocks to optimize the resource sharing
+  // between different streams.
+  // Each stream maintain following data structure:
+  // 1. Current timestamp
+  // 2. A lookup table that for a given stream, what is its timestamp when the
+  //    last synchronization happened with current stream.
+  // 3. When a notificaiton is activated, it take a snapshot of current stream's
+  //    lookup table.
+  // 4. When synchronization happened (current stream wait on a notification),
+  //    update its lookup table with the table snapshot in notificaiton.
+  // The memory reusing strategy is:
+  // A kernel in current stream is safe to reuse another stream's memory chunk
+  // as long as the reused chunk's timestamp is less than the last synchonized
+  // timestamp recorded in the lookup table.
+
+  // Get the current timestamp
+  uint64_t GetCurrentTimestamp() const { return timestamp_; }
+
+  // return the timestamp when the last synchronization happened between target stream and current stream.
+  // return 0 if no synchonization happened.
+  uint64_t GetLastSyncTimestampWithTargetStream(Stream* target_stream) const {
+    auto it = other_stream_clock_.find(target_stream);
+    return it == other_stream_clock_.end() ? 0 : it->second;
+  }
+
+  // make a copy of the current stream lookup table.
+  // this is used to create a snapshot of the stream lookup table in notificaiton.
+  void CloneCurrentStreamSyncTable(std::unordered_map<Stream*, uint64_t>& output) const {
+    std::copy(other_stream_clock_.begin(), other_stream_clock_.end(), std::inserter(output, output.end()));
+  }
+
+  // bump the current timestamp
+  // When a notification get activated, bump the snapshot in its owner.
+  uint64_t BumpTimeStampAndReturn() {
+    return ++timestamp_;
+  }
+
+  // update the stream lookup table with the snapshot saved in notification.
+  void UpdateStreamClock(const std::unordered_map<Stream*, uint64_t>& clock) {
+    for (const auto& kv : clock) {
+      auto it = other_stream_clock_.find(kv.first);
+      if (it == other_stream_clock_.end()) {
+        other_stream_clock_.insert(kv);
+      } else {
+        other_stream_clock_[kv.first] = std::max(it->second, kv.second);
+      }
+    }
+  }
+
+ protected:
+  StreamHandle handle_;
+  const OrtDevice& device_;
+  uint64_t timestamp_{0};
+  // TODO: use inline container.
+  // currently this class is header only, but abseil doesn't compile with nvcc
+  // we need to add new symbol to provider_bridge and hide abseil from the header.
+  std::unordered_map<Stream*, uint64_t> other_stream_clock_{};
+};
+
+namespace synchronize {
+// an object which record the status of the stream, and can be wait on from another stream.
+class Notification {
+ public:
+  explicit Notification(Stream* s) : stream_(s) {}
```

Comment:
can stream_ be null? ActivateAndUpdate() doesn't check before using it.
if not, maybe use `gsl::not_null<Stream*>` or `Stream&` as an argument instead?

Reply:
yes.